Dodge HaxeFoundation/Haxe#11338 by also looking at imported types

kLabz opened the pull request:
See HaxeFoundation/Haxe#11338

kLabz committed:
Dodge HaxeFoundation/Haxe#11338 by also looking at imported types

kLabz committed:
Try locally available types, but still following components resolution rules

kLabz committed:
Should not need to resolve again

yuxiaomao commented:
Seems good now for Shiro's projects.

kLabz commented:
What is the minimum Haxe version supported by domkit? This PR uses a Haxe 4.3.0 feature (`??` operator)

Simn commented:
The problem is that domkit abused https://github.com/HaxeFoundation/haxe/issues/9150 by importing `some.ComponentCollection` to make its types available as `some.Component`. This no longer works after https://github.com/HaxeFoundation/haxe/pull/11338, so a different solution is needed.

kLabz commented:
Can repro the initial issue with dune:

```
 ERROR  src/ui/GameUI.hx:58: characters 4-21

 58 |   <spectator-top-bar/>
    |    ^^^^^^^^^^^^^^^^^
    | Could not load component 'spectator-top-bar'
```

Init macros add this as components path:

```haxe
domkit.Macros.registerComponentsPath("ui.hud.$UI");
```

Module `src/ui/hud/SpectatorUI.hx` also defines component `SpectatorTopBarUI` which we want to resolve here. This used to work because there is `import ui.hud.SpectatorUI` in `GameUI.hx`, so `SpectatorUI` module is loaded and `SpectatorTopBarUI` could be resolved through `ui.hud.SpectatorTopBarUI`.

We cannot resolve `ui.hud.SpectatorTopBarUI` directly anymore, but `SpectatorTopBarUI` is available locally thanks to the import. This is what `Context.resolveComplexType` handles here (it also compares packages to be sure this is not resolving some unrelated component that would have the same name). 

---

There is also this case:

```haxe
domkit.Macros.registerComponentsPath("h2d.domkit.BaseComponents.$Comp");
```

For which I added the `name` vs `sub` handling.

---

As to why I replaced `Context.getType(path)` with `Context.resolveType(TPath(tpath), pos);` there.. I'm not sure anymore tbh, my current guess is that it's mostly a remnant of the early stages of that PR, where I was manipulating a type path instead of a string. I think the two actual differences there are `ParamSpawnMonos` vs `ParamNormal` and exceptions being `String` vs `Dynamic` (but it was already catching `Dynamic` anyway).

ncannasse commented:
@kLabz thanks for the description of the issue in relation to the actual code involved.
However I think this should be resolved by having `"$"` component path registered so when we have an import we can load its components using the name itself. In that case we wouldn't need to change the domkit code at all ?
Can you confirm that @yuxiaomao ?

kLabz commented:
That wouldn't work here either because `$` would resolve to `SpectatorTopBar`, not `SpectatorTopBarUI`

ncannasse commented:
Yeah I meant "$UI"

Le sam. 18 janv. 2025, 13:24, Rudy Ges ***@***.***> a écrit :

> That wouldn't work here either because $ would resolve to SpectatorTopBar,
> not SpectatorTopBarUI
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/HeapsIO/domkit/pull/62#issuecomment-2599695905>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHZXQHTJ4MNRONVBI3TFU32LJBWZAVCNFSM6AAAAABS5MNWY6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGY4TKOJQGU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

kLabz commented:
Seems to work indeed for dune :+1:
